File: src/main/java/kr/hahaha98757/zombiesaddon/features/SpawnLimitAction.java
//Code in Zombies Utils by Stachelbeere1248

package kr.hahaha98757.zombiesaddon.features;

import kr.hahaha98757.zombiesaddon.config.ZombiesAddonConfig;
import kr.hahaha98757.zombiesaddon.data.spawnlimitaction.Room;
import kr.hahaha98757.zombiesaddon.data.spawnlimitaction.Window;
import kr.hahaha98757.zombiesaddon.enums.Map;
import kr.hahaha98757.zombiesaddon.events.TitleEvent;
import kr.hahaha98757.zombiesaddon.utils.Utils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityPlayerSP;
import net.minecraft.client.gui.FontRenderer;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

import java.util.Arrays;

public class SpawnLimitAction {
    private static final int[] offset = new int[3];
    private static Room[] rooms;
    private static int rotations;
    private static boolean isMirroringX;
    private static boolean isMirroringZ;

    public static void setMap(Map map) {
        switch (map) {
            case DEAD_END:
                rooms = Room.getDE();
                break;
            case BAD_BLOOD:
                rooms = Room.getBB();
                break;
            case ALIEN_ARCADIUM:
                rooms = Room.getAA();
                break;
        }
    }

    public static void offMap() {
        rooms = null;
        resetOffset();
    }

    public static void resetRotate() {
        rotate(4 - rotations);
    }

    public static void rotate(int rotations) {
        SpawnLimitAction.rotations = (SpawnLimitAction.rotations + rotations) % 4;
        for (Room room : rooms) for (Window window : room.getWindows()) window.rotate(rotations);
    }

    public static void resetMirroring() {
        if (isMirroringX) mirrorX();
        if (isMirroringZ) mirrorZ();
    }

    public static void mirrorX() {
        isMirroringX = !isMirroringX;
        for (Room room : rooms) for (Window window : room.getWindows()) window.mirrorX();
    }

    public static void mirrorZ() {
        isMirroringZ = !isMirroringZ;
        for (Room room : rooms) for (Window window : room.getWindows()) window.mirrorZ();
    }

    private static void refreshActives() {
        final EntityPlayerSP player = Minecraft.getMinecraft().thePlayer;
        final double[] playerCoords = {
                (player.posX - offset[0]),
                player.posY - offset[1],
                player.posZ - offset[2]
        };
        for (Room room : rooms) {
            room.resetActiveWindowCount();
            for (Window window : room.getWindows()) {
                double distanceDoubledThenSquared = 0;
                for (int i = 0; i < 3; i++)
                    distanceDoubledThenSquared += ((playerCoords[i] * 2 - window.getXYZ()[i]) * (playerCoords[i] * 2 - window.getXYZ()[i]));

                // (2x)²+(2y)²+(2z)² = 4(x²+y²+z²) = 4d²
                final int slaRange = 50;
                if (distanceDoubledThenSquared < 4 * slaRange * slaRange) {
                    window.setActive(true);
                    room.increaseActiveWindowCount();
                } else window.setActive(false);
            }
        }
    }

    public static Room[] getRooms() {
        refreshActives();
        return rooms;
    }

    public static void resetOffset() {
        Arrays.fill(offset, 0);
    }

    public static void setOffset(int[] offset) {
        System.arraycopy(offset, 0, SpawnLimitAction.offset, 0, 3);
    }

    @SubscribeEvent
    public void onRender(RenderGameOverlayEvent.Post event) {
        if (Utils.isModDisable()) return;
        if (event.type != RenderGameOverlayEvent.ElementType.TEXT) return;

        try {
            Room[] rooms = getRooms();

            FontRenderer fr = Minecraft.getMinecraft().fontRendererObj;
            int y = 0;
            for (Room room : rooms) {
                if (!ZombiesAddonConfig.isShowInactiveWindows() && room.getActiveWindowCount() == 0) continue;
                fr.drawStringWithShadow(room.getSLAString(), 1, 1 + y * fr.FONT_HEIGHT, 0xFFFFFF);
                y++;
            }
        } catch (Exception ignored) {
        }
    }

    private static boolean autoSLA;
    private static byte tick;

    @SubscribeEvent
    public void onTitle(TitleEvent event) {
        if (Utils.isModDisable()) return;
        if (!ZombiesAddonConfig.isAutoSLA()) return;

        String title = event.getTitle();

        if (!(Utils.isRoundText(title) && title.replaceAll("[^0-9]", "").equals("1"))) return;

        tick = 0;
        autoSLA = true;
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        if (event.phase != TickEvent.Phase.START) return;

        if (!autoSLA) return;

        Map map = Utils.getMap();

        if (map == null) {
            tick++;
            return;
        }

        if (tick == 100) {
            Utils.addTranslationChat("zombiesaddon.features.sla.auto.wrongMap", new Object());
            autoSLA = false;
            tick = 0;
            return;
        }

        switch (map) {
            case DEAD_END:
                SpawnLimitAction.setMap(Map.DEAD_END);
                Utils.addTranslationChat("zombiesaddon.features.sla.auto.setMap", "§aDead End");
                autoSLA = false;
                tick = 0;
                return;
            case BAD_BLOOD:
                SpawnLimitAction.setMap(Map.BAD_BLOOD);
                Utils.addTranslationChat("zombiesaddon.features.sla.auto.setMap", "§aBad Blood");
                autoSLA = false;
                tick = 0;
                return;
            case ALIEN_ARCADIUM:
                SpawnLimitAction.setMap(Map.ALIEN_ARCADIUM);
                Utils.addTranslationChat("zombiesaddon.features.sla.auto.setMap", "§aAlien Arcadium");
                autoSLA = false;
                tick = 0;
                return;
        }
        tick++;
    }
}

File: src/main/kotlin/kr/hahaha98757/zombiesaddon/modules/SpawnLimitAction.kt
package kr.hahaha98757.zombiesaddon.modules

import kr.hahaha98757.zombiesaddon.ZombiesAddon
import kr.hahaha98757.zombiesaddon.data.Room
import kr.hahaha98757.zombiesaddon.enums.ZombiesMap
import kr.hahaha98757.zombiesaddon.events.GameEndEvent
import kr.hahaha98757.zombiesaddon.events.RoundStartEvent
import kr.hahaha98757.zombiesaddon.utils.JsonLoader
import kr.hahaha98757.zombiesaddon.utils.addTranslatedChat
import kr.hahaha98757.zombiesaddon.utils.fr
import kr.hahaha98757.zombiesaddon.utils.mc
import net.minecraftforge.client.event.RenderGameOverlayEvent
import kotlin.math.pow

object SlaHandler: AlwaysEnableModule("Spawn Limit Action") {
    override fun onRender(event: RenderGameOverlayEvent.Text) {
        Sla.sla?.refresh() ?: return
        val rooms = Sla.sla?.rooms ?: return
        var y = 0
        for (room in rooms) {
            if (!ZombiesAddon.instance.config.slaUnactivatedWindows && room.activeWindows == 0) continue
            fr.drawStringWithShadow(room.toString(), 1f, 1f + fr.FONT_HEIGHT * y++, 0xffffff)
        }
    }

    override fun onRoundStart(event: RoundStartEvent) {
        if (!ZombiesAddon.instance.config.slaAutoSla) return
        if (event.game.round != 1) return
        val map = event.game.gameMode.map
        runCatching { Sla.on(map) }
    }

    override fun onGameEnd(event: GameEndEvent) {
        if (!ZombiesAddon.instance.config.slaAutoSla) return
        Sla.off()
    }
}

class Sla(map: ZombiesMap) {
    companion object {
        var sla: Sla? = null
            private set

        fun off() {
            sla = null
            addTranslatedChat("zombiesaddon.commands.sla.setMap", "§coff")
        }

        fun on(map: ZombiesMap) {
            sla = Sla(map)
            addTranslatedChat("zombiesaddon.commands.sla.setMap", "§a$map")
        }
    }

    val rooms = when (map) {
        ZombiesMap.DEAD_END -> JsonLoader.loadJsonFromResource("data/sla/DE.json", Array<Room>::class.java)
        ZombiesMap.BAD_BLOOD -> JsonLoader.loadJsonFromResource("data/sla/BB.json", Array<Room>::class.java)
        ZombiesMap.ALIEN_ARCADIUM -> JsonLoader.loadJsonFromResource("data/sla/AA.json", Array<Room>::class.java)
        else -> throw IllegalArgumentException("맵 ${map}은(는) SLA를 지원하지 않습니다.")
    }
    private val offset = IntArray(3)
    private var rotations = 0
    private var mirroring = booleanArrayOf(false, false)

    fun resetRotate() {
        rotate(4 - rotations, true)
        addTranslatedChat("zombiesaddon.commands.sla.resetRotations")
    }

    fun rotate(rotations: Int, reset: Boolean = false) {
        var newRotations = rotations % 4
        if (newRotations < 0) newRotations += 4
        sla?.run {
            this.rotations = (this.rotations + newRotations) % 4
            for (room in rooms) for (window in room.windows) window.rotate(newRotations)
            if (reset) return
            addTranslatedChat("zombiesaddon.commands.sla.rotates", newRotations * 90)
        }
    }

    fun resetMirroring() {
        if (mirroring[0]) mirrorX(true)
        if (mirroring[1]) mirrorZ(true)
        addTranslatedChat("zombiesaddon.commands.sla.resetMirroring")
    }

    fun mirrorX(reset: Boolean = false) {
        mirroring[0] = !mirroring[0]
        for (room in rooms) for (window in room.windows) window.mirrorX()
        if (reset) return
        addTranslatedChat("zombiesaddon.commands.sla.mirror", "0 y z")
    }

    fun mirrorZ(reset: Boolean = false) {
        mirroring[1] = !mirroring[1]
        for (room in rooms) for (window in room.windows) window.mirrorZ()
        if (reset) return
        addTranslatedChat("zombiesaddon.commands.sla.mirror", "0 x z")
    }

    fun resetOffset() = setOffset(0, 0, 0)

    fun setOffset(x: Int, y: Int, z: Int) {
        offset[0] = x
        offset[1] = y
        offset[2] = z
        addTranslatedChat("zombiesaddon.commands.sla.setOffset", "$x $y $z")
    }

    fun refresh() {
        val player = mc.thePlayer ?: return
        val playerCoords = doubleArrayOf((player.posX - offset[0]), (player.posY - offset[1]), (player.posZ - offset[2]))
        for (room in rooms) {
            room.activeWindows = 0
            for (window in room.windows) {
                /*
                창문의 좌표는 정수로 표기하기 위해 2배하여 저장되어 있음.
                따라서 플레이어의 좌표도 2배하여 비교해야 함.
                distance는 (2x)^2 + (2y)^2 + (2z)^2 = 4(x^2 + y^2 + z^2) 이다.
                SLA 범위는 50m이므로 4*50^2 = 10000이다.
                이 둘을 비교하여 SLA 범위에 있는지 확인한다.
                 */
                var distance = 0.0
                for (i in 0..2) distance += (playerCoords[i]*2 - window.xyz[i]).pow(2)
                if (distance < 10000) {
                    window.active = true
                    room.activeWindows++
                } else window.active = false
            }
        }
    }
}